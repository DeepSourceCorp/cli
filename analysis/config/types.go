package config

import "github.com/deepsourcelabs/cli/analysis/lsp"

///////////////////////////////////////////////////////
// DSConfig is the struct for .deepsource.toml file //
/////////////////////////////////////////////////////

type DSConfig struct {
	Version         int           `mapstructure:"version,omitempty" json:"version" toml:"version"`
	ExcludePatterns []string      `mapstructure:"exclude_patterns,omitempty" json:"exclude_patterns,omitempty" toml:"exclude_patterns"`
	TestPatterns    []string      `mapstructure:"test_patterns,omitempty" json:"test_patterns,omitempty" toml:"test_patterns"`
	Analyzers       []Analyzer    `mapstructure:"analyzers,omitempty" json:"analyzers,omitempty" toml:"analyzers"`
	Transformers    []Transformer `mapstructure:"transformers,omitempty" json:"transformers,omitempty" toml:"transformers"`
}

type Analyzer struct {
	Name                string      `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	RuntimeVersion      string      `mapstructure:"runtime_version,omitempty" json:"runtime_version,omitempty" toml:"runtime_version,omitempty"`
	Enabled             bool        `mapstructure:"enabled,omitempty" json:"enabled" toml:"enabled"`
	DependencyFilePaths []string    `mapstructure:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `mapstructure:"meta,omitempty" json:"meta,omitempty" toml:"meta"`
	Thresholds          interface{} `mapstructure:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	Enabled bool   `mapstructure:"enabled,omitempty" json:"enabled,omitempty" toml:"enabled"`
}

///////////////////////////////////////
// LSP based Analysis Config Types  //
/////////////////////////////////////

type AnalysisConfig struct {
	Files           []lsp.TextDocumentItem `json:"files"`
	TestFiles       []lsp.TextDocumentItem `json:"test_files"`
	ExcludedFiles   []lsp.TextDocumentItem `json:"excluded_files"`
	ExcludePatterns []string               `json:"exclude_patterns"`
	TestPatterns    []string               `json:"test_patterns"`
	AnalyzerMeta    interface{}            `json:"analyzer_meta"`
}

//////////////////////////////////////
// LSP based Analysis Result Types  //
/////////////////////////////////////

type Namespace struct {
	Key   string  `json:"key"`
	Value float64 `json:"value"`
}

type Metric struct {
	MetricCode string      `json:"metric_code"`
	Namespaces []Namespace `json:"namespaces"`
}

type AnalysisResult struct {
	Issues   []lsp.Diagnostic `json:"issues"`
	Metrics  []Metric         `json:"metrics,omitempty"`
	IsPassed bool             `json:"is_passed"`
	Errors   []Error          `json:"errors"`
	// Errors    []lsp.Diagnostic `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}

////////////////////////////////////
// Default Analysis Result Types  //
///////////////////////////////////

type Error struct {
	HMessage string `json:"hmessage"`
	Level    int    `json:"level"`
}

type Coordinate struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type Position struct {
	Begin Coordinate `json:"begin"`
	End   Coordinate `json:"end"`
}

type Location struct {
	Path     string   `json:"path"`
	Position Position `json:"position"`
}

type Issue struct {
	Code     string   `json:"issue_code"`
	Title    string   `json:"issue_text"`
	Location Location `json:"location"`
}

type DefaultAnalysisResult struct {
	Issues    []Issue     `json:"issues"`
	Metrics   []Metric    `json:"metrics,omitempty"`
	IsPassed  bool        `json:"is_passed"`
	Errors    []Error     `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}
