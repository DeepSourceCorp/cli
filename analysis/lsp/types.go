package lsp

//////////////////////////////
// Document identity types  //
/////////////////////////////

type DocumentURI string

type TextDocumentItem struct {
	URI        DocumentURI `json:"uri"`
	LanguageID string      `json:"languageID,omitempty"`
	Version    int         `json:"version,omitempty"`
	Text       string      `json:"text,omitempty"`
}

///////////////////////////////
// Diagnostic related types //
/////////////////////////////

type DiagnosticSeverity int

const (
	Error       DiagnosticSeverity = 1
	Warning     DiagnosticSeverity = 2
	Information DiagnosticSeverity = 3
	Hint        DiagnosticSeverity = 4
)

type Position struct {
	Line      int `json:"line"`
	Character int `json:"character"`
}
type Range struct {
	Start Position `json:"start"`
	End   Position `json:"end"`
}

type Location struct {
	URI   string `json:"uri"`
	Range Range  `json:"range"`
}

type DiagnosticRelatedInformation struct {
	Location Location `json:"location"`
	Message  string   `json:"message"`
}

type Diagnostic struct {
	Range    Range              `json:"range"`
	Severity DiagnosticSeverity `json:"severity,omitempty"`
	Code     string             `json:"code,omitempty"`
	Source   string             `json:"source,omitempty"`
	Message  string             `json:"message"`

	/**
	 * An array of related diagnostic information, e.g. when symbol-names within
	 * a scope collide all definitions can be marked via this property.
	 * var a,b
	 * a := 2
	 * Issues in line 1 and 2 are related.
	 */
	RelatedInformation []DiagnosticRelatedInformation `json:"relatedInformation"`
}
