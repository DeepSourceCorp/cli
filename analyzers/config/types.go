package config

type Analysis struct {
	Command string `toml:"command" json:"command" validate:"required"`
}

// TODO: Remove the comment when we start support for Autofix.
// type Autofix struct {
//     Command string `toml:"command" json:"command" analyzertoml:"required"`
// }

type Build struct {
	Builder    string `toml:"builder" json:"builder" validate:"omitempty"`
	Dockerfile string `toml:"dockerfile" json:"dockerfile,omitempty" validate:"omitempty"`
	Script     string `toml:"script" json:"script" validate:"omitempty"`
}

type URL struct {
	Source        string `toml:"source" json:"source" validate:"omitempty,url"`
	Documentation string `toml:"documentation" json:"documentation,omitempty" validate:"omitempty,url"`
	BugTracker    string `toml:"bug_tracker" json:"bug_tracker" validate:"omitempty,url"`
}

type Test struct {
	Command string `toml:"command" json:"command" validate:"omitempty"`
}

// analyzer.toml type
type AnalyzerMetadata struct {
	// Analyzer specific data
	Name                 string            `toml:"name" json:"name" validate:"required"`
	Shortcode            string            `toml:"shortcode" json:"shortcode" validate:"required"` // New
	Description          string            `toml:"description" json:"description" validate:"required"`
	Category             string            `toml:"category" json:"category" validate:"required"`
	Tags                 []string          `toml:"tags" json:"tags,omitempty" validate:"omitempty"`
	URL                  URL               `toml:"urls" json:"urls,omitempty" validate:"omitempty"`
	EnvironmentVariables map[string]string `toml:"environment_variables" json:"environment_variables,omitempty" validate:"omitempty"`

	// Analyzer, Autofix and Transformer config
	Analysis Analysis `toml:"analysis" json:"analysis" validate:"required"`
	// Autofix Autofix `toml:"autofix" json:"autofix" validate:"omitempty"`

	// Build steps needed to build the analyzer
	Build Build `toml:"build" json:"build" validate:"omitempty"`

	// Test command for running the testing workflow
	Test Test `toml:"test" json:"test" validate:"omitempty"`
}

// Analyzer issue description
type AnalyzerIssue struct {
	Shortcode   string `validate:"omitempty"`
	Title       string `toml:"title" json:"title" validate:"required"`
	Description string `toml:"description" json:"description" validate:"required"`
	Category    string `toml:"category" json:"category" validate:"required"`
}
