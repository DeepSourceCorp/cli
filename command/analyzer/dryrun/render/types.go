package render

import (
	"html/template"
	"time"

	"github.com/deepsourcelabs/cli/types"
	"github.com/labstack/echo/v4"
)

type RunSummary struct {
	RunDuration       string // Time taken to complete analysis.
	TimeSinceRun      string // Time elapsed since the completion of the analysis run.
	AnalysisStartTime time.Time
	AnalysisEndTime   time.Time
}

type VCSInfo struct {
	Branch      string // VCS branch of the Analyzer.
	CommitSHA   string // The latest commit SHA of the Analyzer.
	VersionDiff string // The string specifying the status of Analyzer w.r.t previous version.
}

type OccurenceData struct {
	IssueMeta  types.AnalyzerIssue // Contains the data stored in issue TOMLs for the respective issue.
	Files      []string            // Files where this issue has been reported.
	FilesInfo  string              // The string containing the data of which files the issue has been reported in.
	Occurences []types.Issue       // The slice of occurences for a certain issue code.
}

type ResultData struct {
	UniqueIssuesCount     int                      // The unique issues count.
	TotalOccurences       int                      // Total issues reported by the Analyzer.
	SourcePath            string                   // The path where the source code to be analyzer is stored.
	IssuesOccurenceMap    map[string]OccurenceData // The map of issue code to its occurences data.
	IssueCategoryCountMap map[string]int           // The map of issue category to the count of the issues of that category.
	AnalysisResult        types.AnalysisResult     // The analysis result post running processors.
	MetricsMap            map[string]float64       // The map of metric names to their values.
	RenderedSourceCode    []template.HTML          // The slice containing the source code snippets for each occurence.
}

type ResultRenderOpts struct {
	EchoServer           *echo.Echo         // The Echo server instance to run the renderer server.
	Template             *template.Template // The go template field so that it can be accessible in `route.go` as well.
	PageTitle            string             // The title of the HTML page.
	AnalyzerShortcode    string             // The shortcode of the Analyzer.
	VCSInfo              VCSInfo            // The VCS information of the Analyzer.
	Summary              RunSummary         // The run summary.
	AnalysisResultData   ResultData         // The analysis result data.
	SelectedIssueCode    string             // The field used to recognize which issue code the user has clicked on to check its occurences.
	SelectedCategory     string             // The field used to recognize which category the user has clicked to filter the issues based on it.
	IssueCategoryNameMap map[string]string  // The map used to route category names to their codes. Eg: `Documentation`->`doc`.
	MetricNameMap        map[string]string  // The map of metrics shortcodes with their names.
}
