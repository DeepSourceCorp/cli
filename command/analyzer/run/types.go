package run

type AnalysisConfig struct {
	Files           []string    `json:"files"`
	ExcludePatterns []string    `json:"exclude_patterns"`
	ExcludeFiles    []string    `json:"exclude_files"`
	TestFiles       []string    `json:"test_files"`
	TestPatterns    []string    `json:"test_patterns"`
	AnalyzerMeta    interface{} `json:"analyzer_meta"`
}
