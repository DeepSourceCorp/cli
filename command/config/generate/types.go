package generate

// DSConfig is the struct for .deepsource.toml file
type Analyzer struct {
	Name                string      `toml:"name,omitempty" json:"name,omitempty"`
	RuntimeVersion      string      `toml:"runtime_version,omitempty" json:"runtime_version,omitempty"`
	Enabled             bool        `toml:"enabled" json:"enabled"`
	DependencyFilePaths []string    `toml:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `toml:"meta,omitempty" json:"meta,omitempty"`
	Thresholds          interface{} `toml:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `toml:"name" json:"name"`
	Enabled bool   `toml:"enabled" json:"enabled"`
}

type DSConfig struct {
	Version         int           `toml:"version" json:"version"`
	ExcludePatterns []string      `toml:"exclude_patterns" json:"exclude_patterns,omitempty"`
	TestPatterns    []string      `toml:"test_patterns" json:"test_patterns,omitempty"`
	Analyzers       []Analyzer    `toml:"analyzers,omitempty" json:"analyzers,omitempty"`
	Transformers    []Transformer `toml:"transformers,omitempty" json:"transformers,omitempty"`
}
