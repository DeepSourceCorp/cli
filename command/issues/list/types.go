package list

// custom types for JSON marshaling

type IssueJSON struct {
	Analyzer       string       `json:"analyzer"`
	IssueCode      string       `json:"issue_code"`
	IssueTitle     string       `json:"issue_title"`
	OccurenceTitle string       `json:"occurence_title"`
	IssueCategory  string       `json:"issue_category"`
	Location       LocationJSON `json:"location"`
}

type LocationJSON struct {
	Path     string       `json:"path"`     // The filepath where the issue is reported
	Position PositionJSON `json:"position"` // The position info where the issue is raised
}

type PositionJSON struct {
	Begin LineColumn `json:"begin"` // The line where the code covered under the issue starts
	End   LineColumn `json:"end"`   // The line where the code covered under the issue starts
}

type LineColumn struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}
