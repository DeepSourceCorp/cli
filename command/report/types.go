package report

// ReportQueryInput is the schema for variables of artifacts
// report GraphQL query
type ReportQueryInput struct {
	AccessToken       string      `json:"accessToken"`
	CommitOID         string      `json:"commitOid"`
	ReporterName      string      `json:"reporter"`
	ReporterVersion   string      `json:"reporterVersion"`
	Key               string      `json:"key"`
	Data              string      `json:"data"`
	AnalyzerShortcode string      `json:"analyzer"`
	Metadata          interface{} `json:"metadata,omitempty"`
}

// ReportQueryInput is the structure of artifacts report
// GraphQL query
type ReportQuery struct {
	Query     string `json:"query"`
	Variables struct {
		Input ReportQueryInput `json:"input"`
	} `json:"variables"`
}

// QueryResponse is the response returned by artifacts report
// GraphQL query
type QueryResponse struct {
	Data struct {
		CreateArtifact struct {
			Error string `json:"error"`
			Ok    bool   `json:"ok"`
		} `json:"createArtifact"`
	} `json:"data"`
}
