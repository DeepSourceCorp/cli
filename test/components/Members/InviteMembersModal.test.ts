import { render } from '@testing-library/vue'
import { createLocalVue, shallowMount } from '@vue/test-utils'
import VTooltip from 'v-tooltip'
import Vuex, { Store } from 'vuex'

import { InviteMembersModal } from '~/components/Members'
import { TeamMemberRoleChoices } from '~/types/types'

interface member {
  index: number
  email: string
  isValid: boolean
  modifyAllowed: boolean
  role: TeamMemberRoleChoices
}

interface InviteMembersModalInterface {
  // Data properties
  showModal: boolean
  membersToInvite: member[]

  // Computed properties
  initialMembersToInvites: member[]
  disableSend: boolean
  pendingInvites: number
  seatsLeftCount: number

  // Methods exposed by the component
  areEmailsInvalid: () => boolean
  addMore: () => void
  openInviteModal: () => void
  close: () => void
  resetInviteLink: () => Promise<void>
  sendInvite: () => Promise<void>
  validEmail: (email: string) => boolean

  // Vuex actions
  inviteAll: (args: {
    ownerId: string
    invitees: {
      email: string
      role: string
    }[]
    login: string
    provider: string
    limit: number
    currentPage: number
  }) => Promise<void>
  resetInviteUrl: (args: { ownerId: string }) => Promise<void>
}

describe('[[ InviteMembersModal ]]', () => {
  // Mocks
  const mocks = {
    $config: {
      emailEnabled: true
    },
    $route: {
      params: {
        provider: 'gh',
        owner: 'deepsourcelabs'
      }
    },
    $toast: {
      danger: jest.fn(),
      success: jest.fn()
    },
    $emit: jest.fn(),
    $fetchState: {
      pending: false
    }
  }

  const stubs = {
    CopyButton: true,
    portal: true,
    ZButton: true,
    ZDivider: true,
    ZIcon: true,
    ZInput: true,
    ZModal: true,
    ZOption: true,
    ZSelect: true
  }

  let store = {} as Store<InviteMembersModal>

  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(VTooltip)

  // Helper that returns a new store instance
  const getStoreInstance = (hasOwnerBillingInfo = true) => {
    const options = {
      modules: {
        'team/detail': {
          namespaced: true,
          state: {
            team: {
              id: 'VGVhbTpxemxyeHo=',
              login: 'deepsourcelabs',
              invitationUrl: 'https://deepsource.icu/invitation/5cnHJHRUdbbpts2ZeiBa-20oabC_W_/',
              invites: {
                totalCount: 1,
                edges: [
                  {
                    node: {
                      email: 'test@test.com',
                      createdAt: '2022-05-25T07:26:58.330500+00:00',
                      role: TeamMemberRoleChoices.Contributor
                    }
                  },
                  {
                    node: {
                      email: 'test2@test.com',
                      createdAt: '2022-05-25T07:28:33.183971+00:00',
                      role: TeamMemberRoleChoices.Admin
                    }
                  }
                ]
              }
            }
          }
        },
        'owner/detail': {
          namespaced: true,
          state: {
            owner: {}
          }
        }
      }
    }

    if (hasOwnerBillingInfo) {
      options.modules['owner/detail'].state.owner = {
        billingInfo: {
          seatsTotal: 5,
          seatsUsed: 2
        }
      }
    }

    const store = new Vuex.Store(options)
    return store as Store<InviteMembersModal>
  }

  // Helper that creates a wrapper with the rendered component
  const getInstance = (includeOwnerBillingInfo = true) => {
    // Get a new store instance
    store = getStoreInstance(includeOwnerBillingInfo)

    const { vm } = shallowMount(InviteMembersModal, {
      store,
      mocks,
      stubs,
      localVue
    })
    return vm as unknown as InviteMembersModalInterface
  }

  // Helper to populate the invite list with Emails
  const populateListEmails = (inviteMemberList: member[]) => {
    return inviteMemberList.map((invitee) => ({
      ...invitee,
      email: `test${invitee.index + 1}@test.com` // test1@test.com, test2@test.com
    }))
  }

  test('renders invite members modal with the trigger button', () => {
    // Get a new store instance
    store = getStoreInstance()

    const { html } = render(InviteMembersModal, {
      data() {
        return {
          showModal: true
        }
      },
      mocks,
      stubs,
      store,
      localVue
    })

    expect(html()).toMatchSnapshot()
  })

  test('`resetInviteLink` method successfully invokes the `resetInviteUrl` Vuex action', async () => {
    const instance = getInstance()

    // Mock implementation for the `resetInviteUrl` Vuex action
    instance.resetInviteUrl = jest.fn().mockResolvedValueOnce(undefined)

    // Invoke the `resetInviteLink` method
    await instance.resetInviteLink()

    // Assertion
    expect(mocks.$toast.success).toBeCalledWith('Invitation link successfully reset')
  })

  test('`areEmailsInvalid` method returns `true` if email entry is not available in `membersToInvite`', () => {
    const instance = getInstance()

    instance.membersToInvite = instance.initialMembersToInvites.map((invitee) => ({
      ...invitee,
      isValid: instance.validEmail(invitee.email)
    }))

    // Assertion
    expect(instance.areEmailsInvalid()).toBe(true)
  })

  test('`areEmailsInvalid` method returns `false` only if all emails in the list are valid', () => {
    const instance = getInstance()

    instance.membersToInvite = populateListEmails(instance.initialMembersToInvites)

    // Assertion
    expect(instance.areEmailsInvalid()).toBe(false)
  })

  test('`validEmail` returns `true` for a valid email', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.validEmail('test@gmail.com')).toBe(true)
  })

  test('`validEmail` returns `false` for an invalid email', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.validEmail('test')).toBe(false)
  })

  test('`validEmail` returns `false` for an empty string', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.validEmail('')).toBe(false)
  })

  test('`sendInvite` returns early if an email supplied is invalid', async () => {
    const instance = getInstance()

    instance.areEmailsInvalid = jest.fn().mockImplementationOnce(() => true)

    // Invoke the `sendInvite` method
    await instance.sendInvite()

    // Assertion
    expect(mocks.$toast.danger).toBeCalledWith('Please enter a valid email')
  })

  test('`sendInvite` method succesfully invokes the `inviteAll` Vuex action', async () => {
    const instance = getInstance()

    instance.membersToInvite = populateListEmails(instance.initialMembersToInvites)

    // Mock implementation for the `inviteAll` Vuex action
    instance.inviteAll = jest.fn().mockResolvedValueOnce(undefined)

    // Set spy
    instance.close = jest.fn()

    // Invoke the `sendInvite` method
    await instance.sendInvite()

    // Assertions
    expect(mocks.$emit).toBeCalledWith('inviteSuccess')
    expect(instance.close).toBeCalled()
  })

  test('`sendInvite` method shows a danger toast if the `inviteAll` Vuex action fails', async () => {
    const instance = getInstance()

    // Mock the behavior
    instance.inviteAll = jest.fn().mockRejectedValueOnce(new Error('Test error'))

    // Invoke the `sendInvite` method
    await instance.sendInvite()

    // Assertion
    expect(mocks.$toast.danger).toBeCalled()
  })

  test('`addMore` method appends to the members invite list', () => {
    const instance = getInstance()

    // Simulate the assignment happening in the fetch hook
    instance.membersToInvite = instance.initialMembersToInvites

    // Invoke the `addMore` method
    instance.addMore()

    // Assertion
    expect(instance.membersToInvite).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          index: 4,
          isValid: true,
          modifyAllowed: false,
          role: TeamMemberRoleChoices.Contributor
        })
      ])
    )
  })

  test('`open` method opens the modal', () => {
    const instance = getInstance()

    // Invoke the `open` method
    instance.openInviteModal()

    // Assertion
    expect(instance.showModal).toBe(true)
  })

  test('`close` method updates `membersToInvite` state variable and closes the modal', () => {
    const instance = getInstance()

    // Invoke the `close` method
    instance.close()

    // Assertions
    expect(instance.membersToInvite).toEqual(instance.initialMembersToInvites)
    expect(instance.showModal).toBe(false)
  })

  test('`disableSend` returns `true` if there are empty emails in the `membersToInvite` list', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.disableSend).toBe(true)
  })

  test('`disableSend` returns `false` if there are no empty emails in the `membersToInvite` list', () => {
    const instance = getInstance()

    instance.membersToInvite = populateListEmails(instance.initialMembersToInvites)

    // Assertion
    expect(instance.disableSend).toBe(false)
  })

  test('`pendingInvites` returns the count of pending invites excluding contributor roles', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.pendingInvites).toBe(1)
  })

  test('`seatsLeftCount` returns `0` if owner billing info is not available', () => {
    const instance = getInstance(false)

    // Assertion
    expect(instance.seatsLeftCount).toBe(0)
  })

  test('`seatsLeftCount` returns the seats left information for the owner', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.seatsLeftCount).toBe(2)
  })

  test('`initialMembersToInvites` returns the members to invite based on `seatsLeftCount`', () => {
    const instance = getInstance(false)

    // Assertion
    expect(instance.initialMembersToInvites).toEqual([
      {
        email: '',
        index: 0,
        isValid: true,
        modifyAllowed: false,
        role: TeamMemberRoleChoices.Contributor
      },
      {
        email: '',
        index: 1,
        isValid: true,
        modifyAllowed: false,
        role: TeamMemberRoleChoices.Contributor
      },
      {
        email: '',
        index: 2,
        isValid: true,
        modifyAllowed: false,
        role: TeamMemberRoleChoices.Contributor
      }
    ])
  })

  test('`initialMembersToInvites` returns the members with `modifyAllowed` set to `true` if there are seats available', () => {
    const instance = getInstance()

    // Assertion
    expect(instance.initialMembersToInvites).toEqual([
      {
        email: '',
        index: 0,
        isValid: true,
        modifyAllowed: true,
        role: TeamMemberRoleChoices.Member
      },
      {
        email: '',
        index: 1,
        isValid: true,
        modifyAllowed: true,
        role: TeamMemberRoleChoices.Member
      },
      {
        email: '',
        index: 2,
        isValid: true,
        modifyAllowed: false,
        role: TeamMemberRoleChoices.Contributor
      }
    ])
  })
})
