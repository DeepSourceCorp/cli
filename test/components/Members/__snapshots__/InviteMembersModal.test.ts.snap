// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[ InviteMembersModal ]] renders invite members modal with the trigger button 1`] = `
"<div>
  <zbutton-stub color=\\"\\" iconcolor=\\"\\" label=\\"\\" loadinglabel=\\"\\" as=\\"button\\" justify=\\"center\\" buttontype=\\"primary\\" type=\\"button\\" size=\\"small\\" to=\\"\\" icon=\\"user-plus\\">
    Invite new member
  </zbutton-stub>
  <portal-stub to=\\"modal\\">
    <zmodal-stub title=\\"Invite new members\\" width=\\"base\\" closeafterprimaryaction=\\"true\\" primaryactionlabel=\\"\\" primaryactiontype=\\"primary\\" showheaderborder=\\"true\\">
      <div class=\\"p-4 space-y-4 text-sm font-normal\\">
        <div class=\\"space-y-2\\">
          <div>
            <h3 class=\\"text-base font-medium text-vanilla-100\\">Invite via email</h3>
            <p class=\\"text-sm text-vanilla-400\\">
              Teammates will create an account via the sent email
            </p>
          </div>
          <div class=\\"flex justify-between\\">
            <zbutton-stub color=\\"\\" iconcolor=\\"\\" label=\\"\\" loadinglabel=\\"\\" as=\\"button\\" justify=\\"center\\" buttontype=\\"link\\" type=\\"button\\" size=\\"small\\" to=\\"\\" spacing=\\"px-0\\">+ Add more emails</zbutton-stub>
            <zbutton-stub color=\\"\\" iconcolor=\\"\\" label=\\"\\" loadinglabel=\\"\\" as=\\"button\\" justify=\\"center\\" buttontype=\\"primary\\" type=\\"button\\" size=\\"small\\" disabled=\\"true\\" to=\\"\\" spacing=\\"px-2\\">
              <div class=\\"flex items-center space-x-2\\">
                <zicon-stub icon=\\"send\\" size=\\"small\\" position=\\"\\" color=\\"ink-400\\"></zicon-stub> <span>Send invite</span>
              </div>
            </zbutton-stub>
          </div>
        </div>
        <zdivider-stub direction=\\"horizontal\\" color=\\"ink-200\\"></zdivider-stub>
        <div class=\\"space-y-2\\">
          <div class=\\"flex items-start justify-between\\">
            <div>
              <h3 class=\\"text-base font-medium text-vanilla-100\\">Invite via link</h3>
              <p class=\\"text-sm text-vanilla-400\\">
                Teammates will create an account via the link
              </p>
            </div>
            <div>
              <zbutton-stub color=\\"\\" iconcolor=\\"\\" label=\\"\\" loadinglabel=\\"\\" as=\\"button\\" justify=\\"center\\" buttontype=\\"secondary\\" type=\\"button\\" size=\\"small\\" to=\\"\\" icon=\\"refresh-cw\\">
                Reset Link
              </zbutton-stub>
            </div>
          </div>
          <zinput-stub name=\\"\\" value=\\"https://deepsource.icu/invitation/5cnHJHRUdbbpts2ZeiBa-20oabC_W_/\\" label=\\"\\" type=\\"text\\" placeholder=\\"Enter a value\\" disabled=\\"true\\" debouncedelay=\\"350\\" size=\\"medium\\" backgroundcolor=\\"ink-400\\" showborder=\\"true\\" errorbordercolor=\\"cherry\\"></zinput-stub>
        </div>
      </div>
    </zmodal-stub>
    <!---->
  </portal-stub>
</div>"
`;
